torch/csrc/autograd/custom_function.cpp

```diff
@@ -26,6 +29,149 @@ Variable VariableInfo::zeros(at::OptionalDeviceGuard& device_guard) const {
   }
 }
 
+void _foreach_tensor(
+    const std::function<void (size_t, size_t, const at::Tensor&)>& fn,
+    torch::jit::Stack* stack,
+    size_t stack_start,
+    size_t size) {
+  // Enumerate over tensors in a stack, including ones in TensorLists
+  int idx = 0;
+  for (const auto i : c10::irange(size)) {
+    auto& ivalue = (*stack)[stack_start + i];
+    if (ivalue.isTensor()) {  // true for optional tensor that has value
+      const auto& tensor = ivalue.toTensor();
+      fn(idx, i, tensor);
+      idx++;
+    } if (ivalue.isTensorList()) {
+      for (const auto& tensor : ivalue.toTensorVector()) {
+        fn(idx, i, tensor);
+        idx++;
+      }
+    }
+  }
+}
+
+void autogradNotImplementedFallback(const c10::OperatorHandle& op, DispatchKeySet dispatch_keys, torch::jit::Stack* stack) {
+  // Mimics the logic of a VariableType NotImplemented kernel
+  const auto& schema = op.schema();
+  const auto& op_name = schema.operator_name().name;
+  const auto& arguments = schema.arguments();
+  const auto& returns = schema.returns();
+  const auto num_arguments = arguments.size();
+  const auto num_returns = returns.size();
+  const auto stack_start = stack->size() - num_arguments;
+  const bool grad_mode = GradMode::is_enabled();
+  std::vector<at::Tensor> tensors_requiring_grad;
+
+  bool any_is_inplace = false;
+  for (const auto i : c10::irange(num_arguments)) {
+    const auto& alias_info = arguments[i].alias_info();
+    any_is_inplace |= alias_info.has_value() && alias_info->isWrite();
+  }
+  // Keep track of which outputs are output of in-place modification
+  // so we can rebase_history if necessary
+  std::vector<bool> is_inplace_output;
+  std::vector<bool> is_aliased_output;
+  is_inplace_output.reserve(num_returns);
+  for (const auto i : c10::irange(num_returns)) {
+    const auto& alias_info = returns[i].alias_info();
+    is_inplace_output.push_back(alias_info.has_value() && alias_info->isWrite());
+    is_aliased_output.push_back(alias_info.has_value());
+  }
+  size_t num_tensor_inputs = 0;  // Only used for DEBUG-only checks
+  for (const auto i : c10::irange(num_arguments)) {
+    auto& ivalue = (*stack)[stack_start + i];
+    if (ivalue.isTensor()) {
+      // ivalues for optional tensors have isTensor = true
+      const auto& tensor = VariableType::unpack(ivalue.toTensor(), op_name.c_str(), i);
+      if (grad_mode && tensor.requires_grad()) {
+        tensors_requiring_grad.push_back(tensor);
+      }
+      num_tensor_inputs++;
+      TORCH_CHECK_NOT_IMPLEMENTED(!generated::details::isFwGradDefined(tensor), "Trying to use forward AD with ", op_name, " that does not support it.");
+    } else if (ivalue.isTensorList()) {
+      const auto& tensors = VariableType::unpack(ivalue.toTensorVector(), op_name.c_str(), i);
```

Comment:
Another bad use of toTensorVector. Here, we have to not use unpack. unpack used to matter when Variable and Tensor were not the same thing but now it is just doing some error checking, and you'll want to just want to do it directly on the ivalue without inducing reference count bump